File: apps/alfred/schemas/notes.py
"""Schemas for Alfred Notes (hierarchical markdown-first pages)."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class NoteCreateRequest(BaseModel):
    title: str | None = Field(default=None, description="Note title (defaults to 'Untitled').")
    icon: str | None = Field(default=None, description="Optional emoji icon.")
    cover_image: str | None = Field(default=None, description="Optional cover image URL.")
    parent_id: str | None = Field(default=None, description="Parent note id (None for root).")
    workspace_id: str | None = Field(default=None, description="Workspace id (defaults to user).")
    content_markdown: str | None = Field(default=None, description="Markdown content.")
    content_json: dict[str, Any] | None = Field(
        default=None, description="Optional editor JSON (e.g., TipTap)."
    )


class NoteUpdateRequest(BaseModel):
    title: str | None = None
    icon: str | None = None
    cover_image: str | None = None
    content_markdown: str | None = None
    content_json: dict[str, Any] | None = None
    is_archived: bool | None = None


class NoteMoveRequest(BaseModel):
    parent_id: str | None = Field(default=None, description="New parent note id (None for root).")
    position: int | None = Field(
        default=None,
        description="New 0-based position within the parent (defaults to end).",
        ge=0,
    )


class NoteSummary(BaseModel):
    id: str
    title: str
    icon: str | None = None
    cover_image: str | None = None
    parent_id: str | None = None
    workspace_id: str
    position: int
    is_archived: bool
    created_at: datetime
    updated_at: datetime


class NoteResponse(NoteSummary):
    content_markdown: str
    content_json: dict[str, Any] | None = None
    created_by: int | None = None
    last_edited_by: int | None = None


class NoteAssetResponse(BaseModel):
    id: str
    note_id: str
    workspace_id: str
    file_name: str
    mime_type: str
    size_bytes: int
    sha256: str | None = None
    url: str
    created_at: datetime
    created_by: int | None = None


class NotesListResponse(BaseModel):
    items: list[NoteSummary]
    total: int
    skip: int
    limit: int


class NoteTreeNode(BaseModel):
    note: NoteSummary
    children: list[NoteTreeNode] = Field(default_factory=list)


class NoteTreeResponse(BaseModel):
    workspace_id: str
    items: list[NoteTreeNode]


class NoteFilesystemEntryResponse(BaseModel):
    name: str
    path: str
    kind: str
    hidden: bool
    importable: bool
    size_bytes: int | None = None
    reason: str | None = None


class NoteFilesystemBrowseResponse(BaseModel):
    path: str
    name: str
    parent_path: str | None = None
    root_path: str
    items: list[NoteFilesystemEntryResponse]


class NoteFilesystemImportRequest(BaseModel):
    workspace_id: str
    path: str
    parent_id: str | None = None
    max_files: int = Field(default=200, ge=1, le=500)


class NoteFilesystemImportResponse(BaseModel):
    source_path: str
    root_note_id: str
    imported_count: int
    skipped_count: int
    skipped_paths: list[str]


__all__ = [
    "NoteCreateRequest",
    "NoteAssetResponse",
    "NoteFilesystemBrowseResponse",
    "NoteFilesystemEntryResponse",
    "NoteFilesystemImportRequest",
    "NoteFilesystemImportResponse",
    "NoteMoveRequest",
    "NoteResponse",
    "NotesListResponse",
    "NoteSummary",
    "NoteTreeNode",
    "NoteTreeResponse",
    "NoteUpdateRequest",
]

File: web/lib/api/types/notes.ts
export type Workspace = {
  id: string;
  name: string;
  icon: string | null;
  user_id: number | null;
  settings: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type WorkspaceCreateRequest = {
  name: string;
  icon?: string | null;
  settings?: Record<string, unknown> | null;
};

export type NoteSummary = {
  id: string;
  title: string;
  icon: string | null;
  cover_image: string | null;
  parent_id: string | null;
  workspace_id: string;
  position: number;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
};

export type NoteResponse = NoteSummary & {
  content_markdown: string;
  content_json: Record<string, unknown> | null;
  created_by: number | null;
  last_edited_by: number | null;
};

export type NoteCreateRequest = {
  title?: string | null;
  icon?: string | null;
  cover_image?: string | null;
  parent_id?: string | null;
  workspace_id?: string | null;
  content_markdown?: string | null;
  content_json?: Record<string, unknown> | null;
};

export type NoteUpdateRequest = {
  title?: string | null;
  icon?: string | null;
  cover_image?: string | null;
  content_markdown?: string | null;
  content_json?: Record<string, unknown> | null;
  is_archived?: boolean | null;
};

export type NotesListResponse = {
  items: NoteSummary[];
  total: number;
  skip: number;
  limit: number;
};

export type NoteTreeNode = {
  note: NoteSummary;
  children: NoteTreeNode[];
};

export type NoteTreeResponse = {
  workspace_id: string;
  items: NoteTreeNode[];
};

export type NoteAssetResponse = {
  id: string;
  note_id: string;
  workspace_id: string;
  file_name: string;
  mime_type: string;
  size_bytes: number;
  sha256: string | null;
  url: string;
  created_at: string;
  created_by: number | null;
};

export type NoteFilesystemEntry = {
  name: string;
  path: string;
  kind: "directory" | "file" | "symlink" | "other";
  hidden: boolean;
  importable: boolean;
  size_bytes: number | null;
  reason: string | null;
};

export type NoteFilesystemBrowseResponse = {
  path: string;
  name: string;
  parent_path: string | null;
  root_path: string;
  items: NoteFilesystemEntry[];
};

export type NoteFilesystemImportRequest = {
  workspace_id: string;
  path: string;
  parent_id?: string | null;
  max_files?: number;
};

export type NoteFilesystemImportResponse = {
  source_path: string;
  root_note_id: string;
  imported_count: number;
  skipped_count: number;
  skipped_paths: string[];
};
